### DISCUSSION TITLE
Host USB serial CDC

### DISCUSSION THREAD

**Discussion by aniplay62** [rsta2/circle#459] [General]
Hi,
If FTDI cable works fine, I tried today CDC with the PI debug probe, but I get a not supported descriptor. I see only a bulk support inside the sources, but the device returns only bmAttributes & 0x3F to 3 and bEndpointAddress to 0x81. I have put a test with bit 0 mask like midi ((bmAttributes & 0x3E) == 0x02) and removed test with (m_pEndpointOut == 0), now I see the device.. but I'am not the best people for continue this job. There are some plans for support more devices ?

**Comment by rsta2** [comment]
My debugprobe device is detected successfully as a serial CDC device here. Do you have a recent [debugprobe firmware](https://github.com/raspberrypi/debugprobe/releases) installed? I haven't tested the data transfer, but it is definitively detected. This is the descriptor dump using `lsusb -v` for the device:

```
Bus 001 Device 003: ID 2e8a:000c Raspberry Pi Debug Probe (CMSIS-DAP)
Device Descriptor:
  bLength                18
  bDescriptorType         1
  bcdUSB               2.10
  bDeviceClass            0 
  bDeviceSubClass         0 
  bDeviceProtocol         0 
  bMaxPacketSize0        64
  idVendor           0x2e8a 
  idProduct          0x000c 
  bcdDevice            1.03
  iManufacturer           1 Raspberry Pi
  iProduct                2 Debug Probe (CMSIS-DAP)
  iSerial                 3 ???
  bNumConfigurations      1
  Configuration Descriptor:
    bLength                 9
    bDescriptorType         2
    wTotalLength       0x0062
    bNumInterfaces          3
    bConfigurationValue     1
    iConfiguration          0 
    bmAttributes         0x80
      (Bus Powered)
    MaxPower              100mA
    Interface Descriptor:
      bLength                 9
      bDescriptorType         4
      bInterfaceNumber        0
      bAlternateSetting       0
      bNumEndpoints           2
      bInterfaceClass       255 Vendor Specific Class
      bInterfaceSubClass      0 
      bInterfaceProtocol      0 
      iInterface              5 CMSIS-DAP v2 Interface
      Endpoint Descriptor:
        bLength                 7
        bDescriptorType         5
        bEndpointAddress     0x04  EP 4 OUT
        bmAttributes            2
          Transfer Type            Bulk
          Synch Type               None
          Usage Type               Data
        wMaxPacketSize     0x0040  1x 64 bytes
        bInterval               0
      Endpoint Descriptor:
        bLength                 7
        bDescriptorType         5
        bEndpointAddress     0x85  EP 5 IN
        bmAttributes            2
          Transfer Type            Bulk
          Synch Type               None
          Usage Type               Data
        wMaxPacketSize     0x0040  1x 64 bytes
        bInterval               0
    Interface Association:
      bLength                 8
      bDescriptorType        11
      bFirstInterface         1
      bInterfaceCount         2
      bFunctionClass          2 Communications
      bFunctionSubClass       2 Abstract (modem)
      bFunctionProtocol       0 
      iFunction               0 
    Interface Descriptor:
      bLength                 9
      bDescriptorType         4
      bInterfaceNumber        1
      bAlternateSetting       0
      bNumEndpoints           1
      bInterfaceClass         2 Communications
      bInterfaceSubClass      2 Abstract (modem)
      bInterfaceProtocol      0 
      iInterface              6 CDC-ACM UART Interface
      CDC Header:
        bcdCDC               1.20
      CDC Call Management:
        bmCapabilities       0x00
        bDataInterface          2
      CDC ACM:
        bmCapabilities       0x02
          line coding and serial state
      CDC Union:
        bMasterInterface        1
        bSlaveInterface         2 
      Endpoint Descriptor:
        bLength                 7
        bDescriptorType         5
        bEndpointAddress     0x81  EP 1 IN
        bmAttributes            3
          Transfer Type            Interrupt
          Synch Type               None
          Usage Type               Data
        wMaxPacketSize     0x0040  1x 64 bytes
        bInterval              16
    Interface Descriptor:
      bLength                 9
      bDescriptorType         4
      bInterfaceNumber        2
      bAlternateSetting       0
      bNumEndpoints           2
      bInterfaceClass        10 CDC Data
      bInterfaceSubClass      0 
      bInterfaceProtocol      0 
      iInterface              0 
      Endpoint Descriptor:
        bLength                 7
        bDescriptorType         5
        bEndpointAddress     0x02  EP 2 OUT
        bmAttributes            2
          Transfer Type            Bulk
          Synch Type               None
          Usage Type               Data
        wMaxPacketSize     0x0040  1x 64 bytes
        bInterval               0
      Endpoint Descriptor:
        bLength                 7
        bDescriptorType         5
        bEndpointAddress     0x83  EP 3 IN
        bmAttributes            2
          Transfer Type            Bulk
          Synch Type               None
          Usage Type               Data
        wMaxPacketSize     0x0040  1x 64 bytes
        bInterval               0
```

As you can see, there are bulk EPs in the CDC data interface (10-0-0).

**Reply by aniplay62** [reply]
**I updated the Pi Debug Probe without success, lsusb -v :**

Bus 003 Device 005: ID 2e8a:000c Raspberry Pi Debug Probe (CMSIS-DAP)
Couldn't open device, some information will be missing
Device Descriptor:
  bLength                18
  bDescriptorType         1
  bcdUSB               2.10
  bDeviceClass            0 
  bDeviceSubClass         0 
  bDeviceProtocol         0 
  bMaxPacketSize0        64
  idVendor           0x2e8a 
  idProduct          0x000c 
  bcdDevice            2.01
  iManufacturer           1 Raspberry Pi
  iProduct                2 Debug Probe (CMSIS-DAP)
  iSerial                 3 E661640843793027
  bNumConfigurations      1
  Configuration Descriptor:
    bLength                 9
    bDescriptorType         2
    wTotalLength       0x0062
    bNumInterfaces          3
    bConfigurationValue     1
    iConfiguration          0 
    bmAttributes         0x80
      (Bus Powered)
    MaxPower              100mA
    Interface Descriptor:
      bLength                 9
      bDescriptorType         4
      bInterfaceNumber        0
      bAlternateSetting       0
      bNumEndpoints           2
      bInterfaceClass       255 Vendor Specific Class
      bInterfaceSubClass      0 
      bInterfaceProtocol      0 
      iInterface              5 
      Endpoint Descriptor:
        bLength                 7
        bDescriptorType         5
        bEndpointAddress     0x04  EP 4 OUT
        bmAttributes            2
          Transfer Type            Bulk
          Synch Type               None
          Usage Type               Data
        wMaxPacketSize     0x0040  1x 64 bytes
        bInterval               0
      Endpoint Descriptor:
        bLength                 7
        bDescriptorType         5
        bEndpointAddress     0x85  EP 5 IN
        bmAttributes            2
          Transfer Type            Bulk
          Synch Type               None
          Usage Type               Data
        wMaxPacketSize     0x0040  1x 64 bytes
        bInterval               0
    Interface Association:
      bLength                 8
      bDescriptorType        11
      bFirstInterface         1
      bInterfaceCount         2
      bFunctionClass          2 Communications
      bFunctionSubClass       2 Abstract (modem)
      bFunctionProtocol       0 
      iFunction               0 
    Interface Descriptor:
      bLength                 9
      bDescriptorType         4
      bInterfaceNumber        1
      bAlternateSetting       0
      bNumEndpoints           1
      bInterfaceClass         2 Communications
      bInterfaceSubClass      2 Abstract (modem)
      bInterfaceProtocol      0 
      iInterface              6 
      CDC Header:
        bcdCDC               1.20
      CDC Call Management:
        bmCapabilities       0x00
        bDataInterface          2
      CDC ACM:
        bmCapabilities       0x06
          sends break
          line coding and serial state
      CDC Union:
        bMasterInterface        1
        bSlaveInterface         2 
      Endpoint Descriptor:
        bLength                 7
        bDescriptorType         5
        bEndpointAddress     0x81  EP 1 IN
        bmAttributes            3
          Transfer Type            Interrupt
          Synch Type               None
          Usage Type               Data
        wMaxPacketSize     0x0040  1x 64 bytes
        bInterval              16
    Interface Descriptor:
      bLength                 9
      bDescriptorType         4
      bInterfaceNumber        2
      bAlternateSetting       0
      bNumEndpoints           2
      bInterfaceClass        10 CDC Data
      bInterfaceSubClass      0 
      bInterfaceProtocol      0 
      iInterface              0 
      Endpoint Descriptor:
        bLength                 7
        bDescriptorType         5
        bEndpointAddress     0x02  EP 2 OUT
        bmAttributes            2
          Transfer Type            Bulk
          Synch Type               None
          Usage Type               Data
        wMaxPacketSize     0x0040  1x 64 bytes
        bInterval               0
      Endpoint Descriptor:
        bLength                 7
        bDescriptorType         5
        bEndpointAddress     0x83  EP 3 IN
        bmAttributes            2
          Transfer Type            Bulk
          Synch Type               None
          Usage Type               Data
        wMaxPacketSize     0x0040  1x 64 bytes
        bInterval               0


**I removed this lines inside GetDescriptor():**
		if (   ucType     == DESCRIPTOR_ENDPOINT
		    && ucDescType == DESCRIPTOR_INTERFACE)
		{
			break;
		}

**Now I see the 2 x Bulk EP descriptors, connected tx to rx, but Tx not works:**

usbdev0-1: Device ven1a40-101, dev9-0-1 found
usbdev0-1: Interface int9-0-0 found
usbdev: Using device/interface int9-0-0
usbdev1-1: Device ven518-1 found
usbdev1-1: Interface int3-1-1 found
usbdev: Using device/interface int3-1-1
usbdev1-1: Interface int3-1-2 found
usbdev: Using device/interface int3-1-2
usbdev1-2: Device ven2e8a-c found
[WARNING] usbdev1-2: Function is not supported
usbdev1-2: Interface int2-2-0 found
usbdev: Using device/interface int2-2-0
[WARNING] umouse: Mouse report size < 3
[DEBUG] usbhub: Port 1: Device configured
GetDescriptor ucDescLen 5, ucDescType 36, ucType 5, 05 24 00 20 01 
GetDescriptor ucDescLen 5, ucDescType 36, ucType 5, 05 24 01 00 02 
GetDescriptor ucDescLen 4, ucDescType 36, ucType 5, 04 24 02 06 
GetDescriptor ucDescLen 5, ucDescType 36, ucType 5, 05 24 06 01 02 
GetDescriptor ucDescLen 7, ucDescType 5, ucType 5, 07 05 81 03 40 00 10 
GetDescriptor ucDescLen 9, ucDescType 4, ucType 5, 09 04 02 00 02 0a 00 00 00 
GetDescriptor ucDescLen 7, ucDescType 5, ucType 5, 07 05 02 02 40 00 00 
GetDescriptor ucDescLen 7, ucDescType 5, ucType 5, 07 05 83 02 40 00 00 
[DEBUG] cdc: Baud rate 9600
[DEBUG] usbhub: Port 2: Device configured
[DEBUG] xhciroot: Port 1: Device configured


Serial test
[DEBUG] cdc: Baud rate 115200

[DEBUG] xhciep: Transfer timed out
[WARNING] userial: USB write failed
[DEBUG] xhciep: Transfer timed out
[WARNING] userial: USB write failed
[DEBUG] xhciep: Transfer timed out
[WARNING] userial: USB write failed
[DEBUG] xhciep: Transfer timed out
[WARNING] userial: USB write failed
[DEBUG] xhciep: Transfer timed out
[WARNING] userial: USB write failed
[DEBUG] xhciep: Transfer timed out
[WARNING] userial: USB write failed
[DEBUG] xhciep: Transfer timed out
[WARNING] userial: USB write failed
[DEBUG] xhciep: Transfer timed out
[WARNING] userial: USB write failed
[DEBUG] xhciep: Transfer timed out
[WARNING] userial: USB write failed
[DEBUG] xhciep: Transfer timed out
[WARNING] userial: USB write failed
[DEBUG] xhciep: Transfer timed out
[WARNING] userial: USB write failed
[DEBUG] xhciep: Transfer timed out
[WARNING] userial: USB write failed

**Comment by rsta2** [comment]
I've found a problem and fixed it on the *develop* branch. How it seems, the Debug Probe UART requires to explicitly enable RTS and DTR to allow data transfers. With this fix the Debug Probe UART is detected and transfers data without problems here with the Circle driver. Thanks for pointing me on this!

I have no explanation for the error messages, you get, and why you have to modify `GetDescriptor()`. If I remember well, you have embedded the Circle USB code into your own system. I can only say, that it works with the native Circle *develop* branch.

**Reply by aniplay62** [reply]
Yes It's USPi, I havn't your latest changes for audio devices but I don't see a difference. Works fine now, both LEDs are flashing. Thanks a lot !

[usbconfigparser.txt](https://github.com/user-attachments/files/16099210/usbconfigparser.txt)

**Reply by rsta2** [reply]
Now I understand it. USPi has differences in the USB configuration parser and cannot handle multiple USB interfaces (functions) in one USB device. Your patch allows the parser to move on to the next interface (CDC data), which contains the bulk endpoints for the data transfer. In Circle this is handled with `SelectInterfaceByClass(10, 0, 0)` in the serial CDC function driver.

**Reply by aniplay62** [reply]
This function is USBFunctionSelectInterfaceByClass() inside USPi, and I call this function : 

[usbserialcdc.txt](https://github.com/user-attachments/files/16100666/usbserialcdc.txt)

And I not see the first interface with EP4 and EP5 :

GetDescriptor ucDescLen 5, ucDescType 36, ucType 5, 05 24 00 20 01
GetDescriptor ucDescLen 5, ucDescType 36, ucType 5, 05 24 01 00 02
GetDescriptor ucDescLen 4, ucDescType 36, ucType 5, 04 24 02 06
GetDescriptor ucDescLen 5, ucDescType 36, ucType 5, 05 24 06 01 02
GetDescriptor ucDescLen 7, ucDescType 5, ucType 5, 07 05 81 03 40 00 10
**GetDescriptor ucDescLen 9, ucDescType 4, ucType 5, 09 04 02 00 02 0a 00 00 00**
GetDescriptor ucDescLen 7, ucDescType 5, ucType 5, 07 05 02 02 40 00 00
GetDescriptor ucDescLen 7, ucDescType 5, ucType 5, 07 05 83 02 40 00 00

With the test not removed, GetDescriptor() stops with the **bold** line and the 2 bulk EP are not found.

**Reply by rsta2** [reply]
Finally, this is your own extension of USPi. USPi support is available in the USPi project, but of course only for functions, which are supported by the project itself.

**Reply by aniplay62** [reply]
Sure i's not easy to update USPI from Circle updates ;-)
Not sure if the function USBFunctionSelectInterfaceByClass() is in the right place, but finally I fixed the problem by a call of USBConfigurationParserCopy() after USBFunctionSelectInterfaceByClass().

**Reply by rsta2** [reply]
Good, that you have found a solution.

**Reply by aniplay62** [reply]
Thanks for your help
